### PR TITLE
Rework with 100% coverage

### PR DESCRIPTION
**Pull request by iproctor**
Consolidate into a single file. Drop vnode store.
Drop the vnode cache idea, just keep a list of old owned vnodes.
Make all the persist callback required.
@jmccarthy14 @jwolski @Raynos @markyen

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,73 +1,82 @@
...
 
-var VNodeStore = require('./lib/vnode-store.js');
+var _ = require('lodash');
+var crypto = require('crypto');
+var async = require('async');
 
-//TODO (joseph@): Config.
-var TOTAL_VNODES = 14;
-var MAX_PARALLEL_TASKS = 5;
+var DEFAULT_TOTAL_VNODES = 14;
```
Comment (jwolski) [inline index.js]: we have to bump this up. it's actually dangerous to defualt this low. would be error prone.

File: index.js [hunk 2]
```diff
@@ -1,73 +1,82 @@
...
+    this.persistKeyToVNode = params.persistKeyToVNode;
+    this.persistRemoveKeyFromVNode = params.persistRemoveKeyFromVNode;
+    this.recoverKeyCallback = params.recoverKey;
+    this.releaseKeyCallback = params.releaseKey;
+    this.totalVNodes = params.totalVNodes || DEFAULT_TOTAL_VNODES;
+    this.logger = params.logger;
+
+    this.ownedVNodes = [];
+
+    this._onRingStateChange(function() {
```
Comment (jwolski) [inline index.js]: why call `onRingStateChange` on instantiation? Just call _attachToRing

File: index.js [hunk 3]
```diff
@@ -1,73 +1,81 @@
...
+    this.hashRingLookup = this.hashRing.lookup.bind(this.hashRing);
+    this.loadVNodeKeysFromStorage = params.loadVNodeKeysFromStorage;
+    this.persistKeyToVNode = params.persistKeyToVNode;
+    this.persistRemoveKeyFromVNode = params.persistRemoveKeyFromVNode;
+    this.recoverKeyCallback = params.recoverKey;
+    this.releaseKeyCallback = params.releaseKey;
+    this.totalVNodes = params.totalVNodes || DEFAULT_TOTAL_VNODES;
+    this.logger = params.logger;
+
+    this.ownedVNodes = [];
```
Comment (jwolski) [inline index.js]: i dont see anything ever added to this guy

Comment (jwolski) [inline index.js]: nvm, i see it now ;)

**Comment by jwolski** [comment]
lgtm. using md5 vs hashcode might be a regression in performance. would like to see a benchmark of them. but general logic looks  for refactor looks sound.

File: test/sevnup.js [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    ring.ready();
+
+    setTimeout(function() {
+        assert.equal(ring.lookup('derp'), 'A');
+        assert.end();
+    });
+});
+
+test('Sevnup._recoverKey handles recoverKey error', function(assert) {
+    var logged = true;
```
Comment (markyen) [inline test/sevnup.js]: `var logged = false`. Also on lines 218 and 240

File: index.js [hunk 5]
```diff
@@ -1,73 +1,81 @@
...
-    this.vnodeStore = new VNodeStore(persistenceService, recoverKey, releaseKey);
+function Sevnup(params) {
+    this.hashRing = params.hashRing;
+    this.hashRingLookup = this.hashRing.lookup.bind(this.hashRing);
+    this.loadVNodeKeysFromStorage = params.loadVNodeKeysFromStorage;
+    this.persistKeyToVNode = params.persistKeyToVNode;
+    this.persistRemoveKeyFromVNode = params.persistRemoveKeyFromVNode;
+    this.recoverKeyCallback = params.recoverKey;
+    this.releaseKeyCallback = params.releaseKey;
+    this.totalVNodes = params.totalVNodes || DEFAULT_TOTAL_VNODES;
```
Comment (jmccarthy14) [inline index.js]: removing the persistence object makes it so you can't swap in specific persistence models, like psql or riak.  needs to be there

**Comment by jmccarthy14** [comment]
+ new tests
+ fixed config numbers to more accurate ringpop
+ mock ringpop
- moved all concerns into one monolithic file (7up file was only concerned with ringpop attachment, events, interfacing with file.  potentially in future taking an enum value for persistence model).
- removed modularity of persistence types (now just mega params)
- removed cache so that per-request-lookups require datastore access

I'm sure you're relying on some of these so i'll come up with commit that keeps them but leaves sep files and vncache (i'll talk to you about that one first) and you can upgrade when you like.

File: index.js (hunk 5 above)
Comment (iproctor) [inline index.js]: just pass in different functions

**Comment by iproctor** [comment]
The cache can be implemented as a wrapper for the persistence functions.

**Pull request merged**